linuxkm/lkcapi_sha_glue.c

```diff
@@ -811,3 +829,238 @@ struct wc_swallow_the_semicolon
+        .cra_ctxsize     =      sizeof(struct wc_linuxkm_drbg_ctx),
+        .cra_init        =      wc_linuxkm_drbg_init_tfm,
+        .cra_exit        =      wc_linuxkm_drbg_exit_tfm,
+        .cra_module      =      THIS_MODULE
+    }
+};
+static int wc_linuxkm_drbg_loaded = 0;
+static int wc_linuxkm_drbg_default_instance_registered = 0;
+
+WC_MAYBE_UNUSED static int wc_linuxkm_drbg_startup(void) {
+    int ret;
+    int cur_refcnt;
+
+    if (wc_linuxkm_drbg_loaded) {
+        pr_err("wc_linuxkm_drbg_set_default called with wc_linuxkm_drbg_loaded.");
+        return -EBUSY;
+    }
+
+    ret = random_test();
+    if (ret) {
+        pr_err("ERROR: self-test for %s failed "
+                           "with return code %d.\n",
+                           wc_linuxkm_drbg.base.cra_driver_name, ret);
+        return -EINVAL;
+    }
+
+    ret = crypto_register_rng(&wc_linuxkm_drbg);
+    if (ret != 0) {
+        pr_err("crypto_register_rng: %d", ret);
+        return ret;
+    }
+
+    wc_linuxkm_drbg_loaded = 1;
+
+    WOLFKM_INSTALL_NOTICE(wc_linuxkm_drbg);
+
+#ifdef LINUXKM_LKCAPI_REGISTER_HASH_DRBG_DEFAULT
+    ret = crypto_del_default_rng();
+    if (ret) {
+        pr_err("crypto_del_default_rng returned %d", ret);
+        return ret;
+    }
+
+    ret = crypto_get_default_rng();
+    if (ret) {
+        pr_err("crypto_get_default_rng returned %d", ret);
+        return ret;
+    }
+
+    cur_refcnt = WC_LKM_REFCOUNT_TO_INT(wc_linuxkm_drbg.base.cra_refcnt);
+    if (cur_refcnt < 2) {
+        pr_err("wc_linuxkm_drbg refcnt = %d after crypto_get_default_rng()", cur_refcnt);
+        crypto_put_default_rng();
+        return -EINVAL;
+    }
+
+    if (! crypto_default_rng) {
+        pr_err("crypto_default_rng is null");
+        crypto_put_default_rng();
+        return -EINVAL;
+    }
+
+    if (strcmp(crypto_tfm_alg_driver_name(&crypto_default_rng->base), wc_linuxkm_drbg.base.cra_driver_name) == 0) {
+        crypto_put_default_rng();
+        wc_linuxkm_drbg_default_instance_registered = 1;
+        pr_info("%s registered as systemwide default stdrng.", wc_linuxkm_drbg.base.cra_driver_name);
+        pr_info("to unload module, first echo 1 > /sys/module/libwolfssl/deinstall_algs");
+    }
+    else {
+        pr_err("%s NOT registered as systemwide default stdrng -- found \"%s\".", wc_linuxkm_drbg.base.cra_driver_name, crypto_tfm_alg_driver_name(&crypto_default_rng->base));
+        crypto_put_default_rng();
+        return -EINVAL;
+    }
+#endif /* LINUXKM_LKCAPI_REGISTER_HASH_DRBG_DEFAULT */
+
+    return 0;
+}
+
+WC_MAYBE_UNUSED static int wc_linuxkm_drbg_cleanup(void) {
+    int cur_refcnt = WC_LKM_REFCOUNT_TO_INT(wc_linuxkm_drbg.base.cra_refcnt);
+    int ret;
+
+    if (! wc_linuxkm_drbg_loaded) {
+        pr_err("wc_linuxkm_drbg_cleanup called with ! wc_linuxkm_drbg_loaded");
+        return -EINVAL;
+    }
+
+    if (cur_refcnt - wc_linuxkm_drbg_default_instance_registered != 1) {
```

Comment:
this can't underflow?

Reply:
Indeed it _can_ underflow if things are terribly messed up, but the test is `!=`, so the bad state would be correctly detected.

Reply:
you're correct, I misread this!